src/python/gudhi/representations/vector_methods.py

```diff
@@ -350,6 +352,154 @@ def __call__(self, diag):
         """
         return self.fit_transform([diag])[0,:]
 
+
+class BettiCurve2(BaseEstimator, TransformerMixin):
+    """
+    A more flexible replacement for the BettiCurve class.
+
+    Examples
+    --------
+    If pd is a persistence diagram and xs is a grid such that xs[0] >= pd.min(), then the result of
+    >>> bc = BettiCurve2(xs)
+    >>> result = bc(pd)
+    and
+    >>> from scipy.interpolate import interp1d
+    >>> bc = BettiCurve2(None)
+    >>> bettis = bc.fit_transform([pd])
+    >>> interp = interp1d(bc.grid_, bettis[0, :], kind="previous", fill_value="extrapolate")
+    >>> result = np.array(interp(xs), dtype=int)
+    are the same.
+    """
+
+    def __init__(self, grid = None):
+        """
+        Constructor for the BettiCurve class.
+
+        Parameters
+        ----------
+        grid: 1d array or None, default=None
+             Filtration grid points at which to compute the Betti curves. Must be strictly ordered. Infinites are OK. If None (default), a grid will be computed that captures all the filtration value changes.
+
+        Attributes
+        ----------
+        grid_: 1d array
+             Contains the compute grid after fit or fit_transform.
+        """
+
+        self.grid_ = np.array(grid)
+
+
+    def fit(self, X, y = None):
+        """
+        Compute a filtration grid that captures all changes in Betti numbers for all the given persistence diagrams.
+
+        Parameters
+        ----------
+        X: list of 2d arrays
+             Persistence diagrams.
+
+        y: None.
+             Ignored.
+        """
+
+        events = np.unique(np.concatenate([pd.flatten() for pd in X], axis=0))
+
+        if len(events) == 0:
+            self.grid_ = np.array([-np.inf])
+        else:
+            self.grid_ = np.array(events)
+
+        return self
+
+
+    def fit_transform(self, X):
+        """
+        Find a sampling grid that captures all changes in Betti numbers, and compute those Betti numbers. The result is the same as fit(X) followed by transform(X), but potentially faster.
+        """
+
+        N = len(X)
+
+        events = np.concatenate([pd.flatten(order="F") for pd in X], axis=0)
+        sorting = np.argsort(events)
+        offsets = np.zeros(1 + N, dtype=int)
+        for i in range(0, N):
+            offsets[i+1] = offsets[i] + 2*X[i].shape[0]
+        starts = offsets[0:N]
+        ends = offsets[1:N + 1] - 1
+
+        bettis = [[0] for i in range(0, N)]
+        if len(sorting) == 0:
```

Comment:
Could you add a very small test for this case? What output do we even want in that case, `np.zeros((N,1))`?

Reply:
That's a good question. I oscillated back and forth over it more than I'll admit. What's your preference?

Reply:
I don't really have one right now, maybe @MathieuCarriere does? I'd be tempted to return an array of shape (N,0), but if that breaks your interpolation identity and zeros((N,1)) doesn't, then that one may be better. This corner case really shouldn't matter much, I asked for a test because you seemed to be going out of your way to support this case (and it is good for the coverage statistics).

Reply:
Yeah, `(N, 0)` sounds like a good option, too. I can just specify that the interpolation identity only holds for non-empty diagrams. I'll wait and see if anyone else has any preferences.

Reply:
I ended up going for the convention that even an empty diagram will return a Betti curve having value `0` from `-np.inf`.

Reply:
I feel like the best would be to return an array whose second dimension is the same as `len(grid)` (if provided). So a list of empty diagrams would return an array full of 0s. If `grid` is not provided, then I personally think that an array of shape `[N,0]` is the most consistent.